### PR TITLE
Use PEP-517 build process for Pipenv, others

### PR DESCRIPTION
**Pull request by danchr**
#### Description

This is related to https://trac.macports.org/ticket/63320 — we should probably use that for more ports, so I tried to enable it for a few of the Python ports I maintain.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS x.y
Xcode x.y

###### Verification 
Have you

- [ ] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ ] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [ ] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by danchr** [comment]
Oh, forgot that I wasn't the maintainer of Dulwich, sorry!

**Comment by lbschenkel** [comment]
It's OK, you can go ahead.

**Comment by reneeotten** [comment]
we'll probably need a change like this at some point, but before making changes to the python PG it should be tested properly on more than a few ports to make sure this will work correctly. This is the suggestion that was made by upstream when I asked about `py-platformdirs` but we have to see whether this works well for MacPorts.

Before potentially merging it we should at least have @jmroot weigh in on the suggested changes.

**Comment by reneeotten** [comment]
@danchr the logs also show `ERROR No module named 'importlib_resources'` so you're likely missing a dependency somewhere; please check the logs for possible additional issues.

File: python/py-build/Portfile [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+maintainers         {danchr @danchr} openmaintainer
+
+description         A simple, correct PEP517 package builder
+long_description    ${description}.
+
+checksums           rmd160  3d76f057048b866a4e162792528c665e98e2be61 \
+                    sha256  16897cac845b50cca04f3c92cf8d3e9e0868b21b29b96b577333c14473baa916 \
+                    size    15243
+
```
Comment (reneeotten) [inline python/py-build/Portfile:25]: missing the `homepage` field

File: python/py-build/Portfile [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+if {${name} ne ${subport}} {
+    depends_build-append \
+                        port:py${python.version}-setuptools \
+                        port:py${python.version}-wheel
+    depends_lib-append  port:py${python.version}-packaging \
+                        port:py${python.version}-pep517 \
+                        port:py${python.version}-toml
+
+    if {${python.branch} < 3.8} {
```
Comment (reneeotten) [inline python/py-build/Portfile:34]: the typical way here would be `if {${python.version} < 38} {`

File: _resources/port1.0/group/python-1.0.tcl [hunk 3]
```diff
@@ -295,12 +295,13 @@ proc python_set_pep517 {option action args} {
         return
     }
     global python.pep517 python.version subport name
-    if {$subport ne $name} {
+    if {[info exists python.version]} {
```
Comment (reneeotten) [inline _resources/port1.0/group/python-1.0.tcl:298]: what's the reason for this change?

Comment (danchr) [reply _resources/port1.0/group/python-1.0.tcl:298]: Without that change, you cannot use PEP-517 builds in single-version ports using `python.version` directly, such as `pipenv`.

File: python/py-build/Portfile (hunk 2 above)
Comment (danchr) [reply python/py-build/Portfile:40]: After a bit of digging, I think it was actually `py-python-install` that was missing a dependency. I've fixed it to use the usual style, thanks.

File: python/py-build/Portfile (hunk 1 above)
Comment (danchr) [reply python/py-build/Portfile:31]: Fixed, thanks!

**Comment by danchr** [comment]
> we'll probably need a change like this at some point, but before making changes to the python PG it should be tested properly on more than a few ports to make sure this will work correctly. This is the suggestion that was made by upstream when I asked about `py-platformdirs` but we have to see whether this works well for MacPorts.

Not many ports use the option, so I've added a commit to rev-bump them all.

**Comment by jmroot** [comment]
OK, I'll first clear up a few points brought up here and in the originating ticket.
 * I am well aware of the evolving python packaging situation.
 * The assertion that `setup.py install` has been obsolete for 5+ years is unfounded. That is when the PEPs were written, but the supporting software, and just as importantly, support in individual projects, didn't appear until much more recently. `platformdirs`, which started this discussion, didn't have a `pyproject.toml` until May this year. Many less maintained projects probably still don't.
 * `pip` is unsuitable for distro packaging generally, because most of what it does either isn't wanted in that context or is redundant and potentially conflicting with what the distro's own tools are doing.
 * We are using `pep517.build` despite its deprecation for similar reasons. Importantly, it has less dependencies than `build`. It seems like `pep517.build` will eventually be split into its own package rather than just being deleted.
 * `setup.py install` works fine in the vast majority of cases.
 * Until a pep517 front end (and preferably also a minimal wheel installer, see above point regarding pip) is included in the standard library, `setup.py install` can't go away without creating severe bootstrap problems. It seems likely that that will come to pass given that `distutils` is deprecated in python 3.10 and there is currently no other way of building the stdlib modules. But for now, there isn't even a TOML parser in the stdlib (partly because the TOML 1.0 spec was only finalised in January this year.)
 * Because of the aforementioned bootstrap problems that are solved by `setup.py install`, no ports that are used as dependencies with `python.pep517` can themselves install with that option. The maintainers of those dependencies have to be vigilant to keep things working if new dependencies are added in new versions.

**Comment by danchr** [comment]
> OK, I'll first clear up a few points brought up here and in the originating ticket.

Thank you for elaborating — those are very valid reasons for not mucking too much about with the PortGroup 🙂 The build also failed with an obscure error in one of the ports, so I've removed the parts of the PR that touched the group. I assume it's still somewhat desirable to use PEP-517 for ports?

**Comment by jmroot** [comment]
Yes, if maintainers want to switch their python module ports to use pep517, that's fine except in the case of the dependencies needed for that build style. That is a direction we should try to be moving in generally. But things do break sometimes, and each port needs to be tested.

File: _resources/port1.0/group/python-1.0.tcl (hunk 3 above)
Comment (jmroot) [reply _resources/port1.0/group/python-1.0.tcl:298]: I'll look at refactoring this. It would probably work better as a callback anyway.

**Comment by jmroot** [comment]
And TBH I would just wait for a version update to do the switch to pep517, rather than rev bumping just for that. But that's up to the maintainers.

I will also look into whether using `build` would gain us significant desirable functionality, and whether that's worth the extra dependencies, and whether the dependencies can be reduced.

**Comment by jmroot** [comment]
I've updated the portgroup to use `build`. The main advantage is that it can check for dependencies and error when they are missing.

The changes to the portgroup and py-python-install in this PR are no longer needed. Enabling pep517 in the other ports should be fine.

**Comment by reneeotten** [comment]
thanks @jmroot, let's just close this PR then...?
